restli-common/src/main/java/com/linkedin/restli/common/validation/RestLiDataValidator.java

```diff
@@ -464,15 +464,15 @@ protected ValidationResult validateOutputAgainstSchema(RecordTemplate dataTempla
    * Checks that if the patch is applied to a valid entity, the modified entity will also be valid.
    * This method
    * (1) Checks that required/ReadOnly/CreateOnly fields are not deleted.
-   * (2) Checks that new values for record templates contain all required fields.
+   * (2) Checks that new values for record templates contain all required fields (treating ReadOnly fields as optional).
    * (3) Applies the patch to an empty entity and validates the entity for custom validation rules
    * and Rest.li annotations (Allows required fields to be absent by using {@link RequiredMode#IGNORE},
    * because a patch does not necessarily contain all fields).
    *
-   * NOTE: Updating a part of an array is not supported. So if the array contains a required field that is
-   * readonly or createonly, the field cannot be present (no partial updates on readonly/createonly)
-   * but cannot be absent either (no missing required fields). This means the array cannot be changed by a
-   * partial update request. This is something that should be fixed.
+   * NOTE: Updating a part of an array or other complex object is not supported if the object contains a required field
+   * that is CreateOnly. In this case, the field cannot be present (no partial updates on CreateOnly fields), but cannot
+   * be absent either (no missing required fields).
+   * TODO: This is something that should be fixed.
```

Comment:
I think what the original comment was trying to say is: "Partial update is not supported on arrays, as we always replace the entire array". 
So in the patch, the entire array record has to be present (full update).

This is not true for other complex types or Maps.  So the original comment is correct and this issue is specific to arrays.
So in partial update, the required createOnly fields must be present (it must be treated as update, so for existing items they should've been present in original and for new items they behave like create) and readonly fields should be treated as optional (can be present for items being updated and can be absent for new array items)

For Maps/other complex types, patch is supported, so if the operation is $set, then they should be treated as CREATE (ie, CreateOnly fields are required, ReadOnly fields should not be present).